### PR TITLE
[inductor] Check dtype in logcumsumexp meta to match eager

### LINKED ISSUE #197807: `torch.compile` (inductor) silently accepts `logcumsumexp` on a 0-d uint8 tensor that eager rejects
### 🐛 Describe the bug

`torch.logcumsumexp` is not implemented for the `uint8` (`Byte`) dtype — eager raises
`NotImplementedError`. Under `torch.compile` (inductor), a **0-dimensional** `uint8` input is
**silently accepted** (returned unchanged) instead of raising, because inductor lowers the
0-d cumulative op to an identity/copy that never reaches the dtype-checked kernel. A 1-D `uint8`
input is correctly rejected by both. So the compiled path drops the dtype validation, but only
on the 0-d fast path.

## Minimal repro

```python
import torch

def fn(a):
    return torch.logcumsumexp(a, dim=0)

x = torch.tensor(0, dtype=torch.uint8)          # 0-d (scalar) uint8

fn(x)                                            # eager:    NotImplementedError ("... not implemented for 'Byte'")
torch.compile(fn, backend="inductor")(x)         # compiled: OK -> tensor(0, dtype=torch.uint8)   (silently accepted)
```

Scope (what does / doesn't diverge):

| input | eager | compiled |
|---|---|---|
| **0-d `uint8`** | NotImplementedError (Byte) | **accepts -> tensor(0, uint8)** |
| 1-D `uint8` | NotImplementedError (Byte) | NotImplementedError (Byte) — consistent |
| 0-d `float32` | OK (`1.5`) | OK — consistent |
| 1-D `float32` | OK | OK — consistent |

So the divergence is specific to **0-d + an unsupported dtype**: inductor's scalar shortcut
skips the dtype check.

## Expected vs actual

- **Expected**: compiled and eager agree — either both reject `uint8` or both accept it.
- **Actual**: eager raises `NotImplementedError` for `Byte`; inductor's 0-d path returns the
  input unchanged, so a program that errors in eager silently "works" once compiled.

### Versions

```
PyTorch version: 2.15.0a0+gite67e43b
Is debug build: True
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04.3) 12.3.0
Clang version: 17.0.6
CMake version: version 4.0.0
Libc version: glibc-2.35

Python version: 3.12.13 | packaged by conda-forge (64-bit runtime)
Python platform: Linux-6.8.0-138-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.6.20
GPU 0: NVIDIA RTX A6000
Nvidia driver version: 580.82.07
cuDNN version: 9.3.0

[pip3] numpy==2.3.5
[pip3] torch==2.15.0a0+gite67e43b
[pip3] triton==3.7.1
```

cc @malfet @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by greekera1000**
Fixes #197807

### Problem
The meta function for `aten.logcumsumexp` only validated `dim`, so tracing accepted
integer/bool inputs that the eager kernels reject. Two symptoms under `torch.compile`:

- **0-d inputs (CPU):** the inductor lowering takes its 0-d shortcut (`clone(x)`), never
  reaches a dtype-checked kernel, and silently returns the input while eager raises
  `NotImplementedError`.
- **1-D inputs (CUDA):** inductor generates a Triton scan kernel for the integer input,
  which fails to compile (`libdevice.isinf` has no integer overload), surfacing as an
  `InductorError` instead of a clean `NotImplementedError`.

### Fix
Add a floating/complex dtype check to the `logcumsumexp` meta function, using
`torch._check_not_implemented` so the error type matches eager.

### Results
Tested on nightly `2.15.0.dev20260907+cu126` (`f2ac313358`), Tesla T4.
Dtypes: uint8, int32, int64, bool.

| Case | Eager | Compiled (before) | Compiled (after) |
|---|---|---|---|
| CPU, 0-d | NotImplementedError | **accepted (silent)** | NotImplementedError |
| CPU, 1-D | NotImplementedError | NotImplementedError | NotImplementedError |
| CUDA, 0-d | accepted | accepted | accepted |
| CUDA, 1-D | NotImplementedError | **InductorError (Triton compile failure)** | NotImplementedError |
| float32/float64 (+ fp16/bf16 on CUDA) | OK | OK, matches eager | OK, matches eager |

Eager/compiled mismatches: **4 before, 0 after**.

Note: eager CUDA accepts 0-d integer inputs while eager CPU rejects them. That's a
pre-existing inconsistency in the eager kernels and out of scope here; compiled
behaviour now matches eager on each device.

### Test plan
Added `test_logcumsumexp_unsupported_dtype_matches_eager` to `CommonTemplate` in
`test/inductor/test_torchinductor.py` (runs on CPU and GPU). It checks that compiled
raises `NotImplementedError` exactly when eager does, for uint8/int32/int64/bool at
0-d and 1-D. It fails without this change (CPU 0-d cases and the CUDA 1-D Triton crash).

```
python test/inductor/test_torchinductor.py -k logcumsumexp
```
disclosure: Drafted with assistance from Claude, I verified the fix on a
Kaggle T4 and reviewed all changes.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by greekera1000** [comment]
Thanks, this is very helpful — you're right that the CUDA 0-d parity was coming from the
FakeTensor unsafe fallback rather than from the meta function, and I should have called
that out.

I'd like to go with the preferred option and add the `TORCH_CHECK_NOT_IMPLEMENTED` to the
`logcumsumexp` / `logcumsumexp_out` composites in `ReduceOps.cpp`, ahead of the CUDA 0-d
and empty fast paths in `_logcumsumexp_out_cuda`. Before I do: are you happy with 0-d and
empty integer tensors on CUDA starting to raise? That's the BC change you mentioned.

I'll also simplify the test to `@parametrize` + `assertRaises`, cover `logcumsumexp.out`
and `dynamic=True`, and add an integer-dtype `ErrorInput` to `error_inputs_logcumsumexp`.

One note: I don't have a machine that can do a full source build, so I'll be relying on CI
to validate the ATen change.
